plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py

```diff
@@ -0,0 +1,226 @@
+        cmd = [
+            ZPOOL_BIN, 'create', '-f',
+            name, '-m', mountpoint
+        ]
+
+        if 'mode' in configuration:
+            if configuration['mode'] not in [
+                    'N', 'M', 'R']:
+                log.error('mode can only be N(default), M(mirror) or R(raidz)')
+                raise
+            if configuration['mode'] in ['M']:
+                if len(devs) < 2:
+                    log.error('mirror mode requires at least two devices')
+                    raise
+                cmd.append('mirror')
+            if configuration['mode'] in ['R']:
+                if len(devs) < 2:
+                    log.error('raidz mode requires at least two devices')
+                    raise
+                cmd.append('raidz')
+
+        cmd.extend(devs)
+
+        try:
+            call(dbg, cmd)
+        except:
+            log.error('error creating the pool')
+            raise
+
+        if compression:
+            cmd = [
+                ZFS_BIN, 'set', 'compression=on', name
+            ]
+            call(dbg, cmd)
+
+        log.debug('{}: SR.create: sr={}'.format(dbg, mountpoint))
+
+        importlib.import_module('zfs-ng').Callbacks().create_database(mountpoint)
+
+        meta = {
+            'name': name,
+            'description': description,
+            'uri': mountpoint,
+            'mountpoint': mountpoint,
+            'unique_id': sr_uuid,
+            'read_caching': False,
+            'keys': {}
+        }
+        util.update_sr_metadata(dbg, 'file://' + mountpoint, meta)
+
+        return configuration
+
+    def destroy(self, dbg, sr):
+        name = os.path.basename(sr)
+        cmd = [
+            ZPOOL_BIN, 'destroy',
+            name
+        ]
+
+        log.debug('cmd={}'.format(cmd))
+
+        try:
+            call(dbg, cmd)
+        except:
+            log.debug('error destroying the pool the pool')
+
+    def detach(self, dbg, sr):
+        # Nothing todo.
+        pass
+
+    def ls(self, dbg, sr):
+        results = []
+        cb = importlib.import_module('zfs-ng').Callbacks()
+        with VolumeContext(cb, sr, 'r') as opq:
+            with cb.db_context(opq) as db:
+                vdis = db.get_all_vdis()
+                all_custom_keys = db.get_all_vdi_custom_keys()
+
+            for vdi in vdis:
+                # TODO: handle this better
+                # _vdi_sanitize(vdi, opq, db, cb)
+
+                image_format = ImageFormat.get_format(vdi.image_type)
+                is_snapshot = bool(vdi.volume.snap)
+                if is_snapshot:
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.parent_id) + '@' + str(vdi.volume.id)
+                else:
+                    path = os.path.basename(sr) + '/'+ str(vdi.volume.id)
+                out = ZFSUtil.get_vsize(dbg, path)
+                if is_snapshot:
+                    psize = int(out[0])
+                else:
+                    psize = int(out[0]) + int(out[1])
```

Comment:
I don't understand this part. :)

Reply:
I was computing wrongly computing the size. I fixed by just using the used value, i.e., out[0], however for a snapshot this is 0 and only increases when snapshot changes from the original volume.